launcher/internal/logging/logging.go

```diff
@@ -0,0 +1,255 @@
+// Package logging implements a logger to be used in the client.
+// Logs to both Cloud Logging and the serial console.
+package logging
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"os"
+
+	"cloud.google.com/go/compute/metadata"
+	clogging "cloud.google.com/go/logging"
+	mrpb "google.golang.org/genproto/googleapis/api/monitoredres"
+)
+
+const (
+	logName           = "confidential-space-launcher"
+	serialConsoleFile = "/dev/console"
+
+	payloadMessageKey  = "MESSAGE"
+	payloadHostnameKey = "_HOSTNAME"
+)
+
+// Logger defines the interface for the CS image logger.
+type Logger interface {
+	Log(severity clogging.Severity, msg string, args ...any)
+
+	Info(msg string, args ...any)
+	Warn(msg string, args ...any)
+	Error(msg string, args ...any)
+
+	SerialConsoleFile() *os.File
+	Close()
+}
+
+type cLogger interface {
+	Log(clogging.Entry)
+	Flush() error
+}
+
+type logger struct {
+	cloudLogger  cLogger
+	serialLogger *slog.Logger
+	resource     *mrpb.MonitoredResource
+
+	instanceName      string
+	cloudClient       *clogging.Client
+	serialConsoleFile *os.File
+}
+
+type payload map[string]any
+
+// NewLogger returns a Logger with Cloud and Serial Console logging configured.
+func NewLogger(ctx context.Context) (Logger, error) {
+	// Retrieve monitored resource information.
+	mdsClient := metadata.NewClient(nil)
+
+	projectID, err := mdsClient.ProjectIDWithContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	instanceID, err := mdsClient.InstanceIDWithContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	instanceName, err := mdsClient.InstanceNameWithContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	zone, err := mdsClient.ZoneWithContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Configure Cloud Logging client/logger.
+	cloggingClient, err := clogging.NewClient(ctx, projectID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Configure Serial Console logger.
+	serialConsole, err := os.OpenFile(serialConsoleFile, os.O_WRONLY, 0)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open serial console for writing: %v", err)
+	}
+
+	slg := slog.New(slog.NewTextHandler(serialConsole, nil))
```

Comment:
Can you show an example log? I assume it doesn't break our cloud bulid tests, but let's make sure they also don't break our e2e tests.

Reply:
Added screenshots for what each type of log will look like in the PR description.